transports/quic/src/endpoint.rs

```diff
@@ -0,0 +1,493 @@
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+//
+use crate::{connection::EndpointMessage, error::Error, socket, Config, Upgrade};
+use async_macros::ready;
+use async_std::net::SocketAddr;
+use futures::{channel::mpsc, prelude::*};
+use libp2p_core::{
+    multiaddr::{Multiaddr, Protocol},
+    transport::{ListenerEvent, TransportError},
+    Transport,
+};
+use log::{debug, info, trace, warn};
+use parking_lot::Mutex;
+use quinn_proto::{Connection, ConnectionHandle};
+use std::{
+    collections::HashMap,
+    pin::Pin,
+    sync::{Arc, Weak},
+    task::{Context, Poll},
+    time::Instant,
+};
+
+#[derive(Debug)]
+pub(super) struct EndpointInner {
+    inner: quinn_proto::Endpoint,
+    muxers: HashMap<ConnectionHandle, Weak<Mutex<super::connection::Muxer>>>,
+    pending: socket::Pending,
+    /// Used to receive events from connections
+    event_receiver: mpsc::Receiver<EndpointMessage>,
+}
+
+impl EndpointInner {
+    pub(super) fn handle_event(
+        &mut self,
+        handle: quinn_proto::ConnectionHandle,
+        event: quinn_proto::EndpointEvent,
+    ) -> Option<quinn_proto::ConnectionEvent> {
+        if event.is_drained() {
+            let res = self.inner.handle_event(handle, event);
+            self.muxers.remove(&handle);
+            res
+        } else {
+            self.inner.handle_event(handle, event)
+        }
+    }
+
+    fn drive_events(&mut self, cx: &mut Context) {
+        while let Poll::Ready(e) = self.event_receiver.poll_next_unpin(cx).map(|e| e.unwrap()) {
+            match e {
+                EndpointMessage::ConnectionAccepted => {
+                    debug!("accepting connection!");
+                    self.inner.accept();
+                }
+                EndpointMessage::EndpointEvent { handle, event } => {
+                    debug!("we have an event from connection {:?}", handle);
+                    match self.muxers.get(&handle).and_then(|e| e.upgrade()) {
+                        None => drop(self.muxers.remove(&handle)),
+                        Some(connection) => {
+                            let event = self.inner.handle_event(handle, event);
+                            connection.lock().process_connection_events(self, event)
+                        }
+                    }
+                }
+                EndpointMessage::Dummy => {}
+            }
+        }
+    }
+
+    fn drive_receive(
+        &mut self,
+        socket: &socket::Socket,
+        cx: &mut Context,
+    ) -> Poll<Result<(ConnectionHandle, Connection), Error>> {
+        use quinn_proto::DatagramEvent;
+        let mut buf = vec![0; 65535];
```

Comment:
This allocates a 64kiB buffer and fills it with 0s every single time we try to receive from the socket :-/